include/revng/TupleTree/TupleTreeDiff.h

```diff
@@ -57,24 +151,139 @@ struct TupleTreeDiff {
     llvm::raw_os_ostream OutputStream(dbg);
     dump(OutputStream);
   }
 
-  void apply(T &M) const;
+  void apply(TupleTree<T> &M) const;
+};
+
+template<typename T>
+concept DiffSpecialization = is_specialization_v<T, TupleTreeDiff>;
+
+template<typename T>
+concept ChangeSpecialization = is_specialization_v<T, Change>;
+
+template<DiffSpecialization T>
+struct llvm::yaml::MappingTraits<T> {
+  static void mapping(IO &IO, T &Info) {
+    IO.mapOptional("changes", Info.Changes);
+  }
+};
+
+template<ChangeSpecialization T, typename X>
+struct llvm::yaml::SequenceElementTraits<T, X> {
+  // NOLINTNEXTLINE
+  static const bool flow = false;
+};
+
+namespace detail {
+template<typename Model>
+struct MapDiffVisitor {
+  llvm::yaml::IO *Io;
+  TupleTreeEntriesT<Model> *Change;
+  const char *MappingName;
+
+  template<typename T, int I>
+  void visitTupleElement() {
+    using tuple_element = typename std::tuple_element<I, T>::type;
+    visit<tuple_element>();
+  }
+
+  template<typename T, typename KeyT>
+  void visitContainerElement(KeyT Key) {}
+
+  template<IsContainer T>
+  void visit() {
+    dump<typename T::value_type>();
+  }
+
+  template<typename T>
+  void visit() {
+    dump<T>();
+  }
+
+  template<typename T>
+  void dump() {
+    if (Io->outputting()) {
+      Io->mapRequired(MappingName, std::get<T>(*Change));
+    } else {
+      T Content;
+      Io->mapRequired(MappingName, Content);
+      *Change = std::move(Content);
+    }
+  }
+};
+
+} // namespace detail
+
+template<ChangeSpecialization T>
+struct llvm::yaml::MappingTraits<T> {
+  using EntryType = typename T::EntryType;
+  using Model = typename T::TupleTreeType;
+
+  static void writeEntry(IO &IO, T &Info, const char *Name, EntryType &Entry) {
+    if (not Entry.has_value())
+      return;
+
+    ::detail::MapDiffVisitor<Model> Visitor{ &IO, &*Entry, Name };
+    callByPath<Model>(Visitor, Info.Path);
+  }
+
+  static void readEntry(IO &IO, T &Info, const char *Name, EntryType &Entry) {
+    const auto &Keys = IO.keys();
+    if (llvm::find(Keys, Name) == Keys.end())
+      return;
+
+    Entry = false;
+    ::detail::MapDiffVisitor<Model> Visitor{ &IO, &*Entry, Name };
+    callByPath<Model>(Visitor, Info.Path);
+  }
+
+  static void
+  mapSingleEntry(IO &IO, T &Info, const char *Name, EntryType &Entry) {
+    if (IO.outputting())
+      writeEntry(IO, Info, Name, Entry);
+    else
+      readEntry(IO, Info, Name, Entry);
+  }
+
+  static void mapping(IO &IO, T &Info) {
+    if (IO.outputting()) {
+      std::string SerializedPath = *pathAsString<Model>(Info.Path);
+      IO.mapRequired("path", SerializedPath);
+    } else {
+      std::string SerializedPath;
+      IO.mapRequired("path", SerializedPath);
+      auto MaybePath = stringAsPath<Model>(SerializedPath);
+      revng_assert(MaybePath.has_value());
+      Info.Path = std::move(*MaybePath);
+    }
+
+    mapSingleEntry(IO, Info, "add", Info.New);
+    mapSingleEntry(IO, Info, "remove", Info.Old);
+  }
```

Comment:
Let's use capital letters for YAML document keys. I'm thinking to switch to lower case eventually, but I prefer coherence for now.